src/commonMain/kotlin/com/mattprecious/stacker/stack/tree.kt

```diff
@@ -0,0 +1,81 @@
+package com.mattprecious.stacker.stack
+
+fun <K : Any, T : Any> treeOf(
+	elements: Collection<T>,
+	keySelector: (T) -> K,
+	parentSelector: (T) -> K?,
+): TreeNode<T>? {
+	if (elements.isEmpty()) {
+		return null
+	}
+
+	val elementsByKey = elements.associateBy(keySelector)
+	val parents = elements.associateBy(keySelector) {
+		elements.indexOf(elementsByKey[parentSelector(it)])
+	}
+	val children = elements.groupBy(parentSelector) { elements.indexOf(it) }
+
+	require(parents.values.count { it == -1 } == 1) {
+		"Multiple elements have a null parent."
+	}
+
+	return Tree(
+		elements = elements.toList(),
```

Comment:
This `toList()` is useless. You immediately call `.map` on it in the `Tree`'s `nodes` initializer which returns a `List`. Pass in the collection directly and save an iteration.